### PR TITLE
Refactor ipac table writer

### PR DESCRIPTION
**Pull request by judith-ipac**
Hi, Serge.   I recommend reviewing these changes one commit at a time.  Most of the commits consist either of renaming a directory or of moving hunks of code between files.  The last non-test commit exposes functions that g2p can use.

Thanks.

### CODE REVIEW

File: src/tablator/main.cxx [hunk 1]
```diff
@@ -52,13 +72,22 @@ int main(int argc, char *argv[]) {
...
+            if (!option_variables.count("row-count")) {
+                row_count = 1;
+            }
+        }
+        if (option_variables.count("row-list")) {
+            std::stringstream row_stream(row_string);
+            std::vector<size_t> temp_list((std::istream_iterator<size_t>(row_stream)),
+                                          std::istream_iterator<size_t>());
+            row_list.swap(temp_list);
+        }
```
Comment (smonkewitz) [inline src/tablator/main.cxx]: Nitpicking, but shouldn't `start-row` and `row-list` be mutually exclusive? And `row-count` shouldn't be allowed without `start-row`, right?

File: src/Data_Type_Adjuster.hxx [hunk 2]
```diff
@@ -10,8 +10,9 @@ public:
     std::vector<tablator::Data_Type> get_datatypes_for_writing(
             const Format::Enums &enum_format) const;
 
-    Data_Type get_datatype_for_writing(
-            const std::vector<Data_Type> &datatypes_for_writing, size_t col) const;
+    static Data_Type get_datatype_for_writing(
+            const Table &table, const std::vector<Data_Type> &datatypes_for_writing,
+            size_t col);
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Data_Type_Adjuster.hxx]: Does this mean you think the sanity check is no longer necessary?

File: src/Ipac_Table_Writer.hxx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <string>
+#include <vector>
+
+#include "Data_Type.hxx"
+
+namespace tablator {
+class Table;
+
+// JTODO Class? Namespace?  It lacks state but does have private functions.
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Ipac_Table_Writer.hxx]: If I could have a namespace with private functions, I'd do that.  I'll change the method names.

File: src/Ipac_Table_Writer/internals.cxx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+void tablator::Ipac_Table_Writer::convert_newlines(std::string& input) {
+    boost::replace_all(input, "\n", "  ");
+}
+
+const std::string tablator::Ipac_Table_Writer::convert_newlines(
+        const std::string& input) {
+    std::string temp_str(input);
+    convert_newlines(temp_str);
+    return temp_str;
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Ipac_Table_Writer/internals.cxx]: I wondered that too. :-)  I inherited the name along with a hunk of code elsewhere accomplishing the same thing, but yes, the boost call is clearer.

File: src/Ipac_Table_Writer/write_ipac_table_header.cxx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+/* Helper functions */
+/*******************************************************/
+
+namespace {
+
+void write_keyword_header_line(std::ostream &os, const std::string &name,
+                               const std::string &value) {
+    os << "\\" << std::setw(KEYWORD_ALIGNMENT)
+       << tablator::Ipac_Table_Writer::convert_newlines(name) << " = "
+       << "'" << tablator::Ipac_Table_Writer::convert_newlines(value) << "'\n";
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Ipac_Table_Writer/write_ipac_table_header.cxx]: My aim was just to preserve the status quo (and only to fix existing bugs if they jumped out at me).

[the remark this answers is not in the document]
Comment (smonkewitz) [reply src/Ipac_Table_Writer/write_ipac_table_header.cxx]: OK, makes sense.

File: src/Data_Type_Adjuster.hxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (smonkewitz) [reply src/Data_Type_Adjuster.hxx]: Yes - if there's no way for an API client to specify datatypes for writing directly.

File: src/Ipac_Table_Writer/internals.cxx [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    write_consecutive_ipac_records(table, os, 0, table.num_rows(),
+                                   datatypes_for_writing);
+}
+
+/**********************************************************/
+
+void tablator::Ipac_Table_Writer::write_ipac_column_headers(
+        const Table& table, std::ostream& os,
+        const std::vector<Data_Type>& datatypes_for_writing) {
+    const std::vector<size_t> ipac_column_widths = table.get_column_widths();
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Ipac_Table_Writer/internals.cxx]: Ooh, yes.

File: src/Ipac_Table_Writer/write_ipac_table_header.cxx [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    os << "\\" << tablator::Table::ROWS_RETRIEVED_KEYWORD << " = " << num_rows_to_report
+       << "\n";
+
+    auto &comments = table.comments;
+    std::vector<std::string> json_comments;
+
+    // Iterate through properties, distinguishing between keywords and json5-ified
+    // comments. Write keywords here and save json_comments for a later loop. (Standard
+    // refers to keywords rather than properties.)
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Ipac_Table_Writer/write_ipac_table_header.cxx]: Out it goes.

File: src/Ipac_Table_Writer.hxx [hunk 8]
```diff
@@ -13,40 +15,79 @@ class Table;
...
+                                        int num_rows = WRITE_ALL_ROWS);
+    static void write_ipac_column_headers(const Table &table, std::ostream &os);
+    static void write_ipac_subtable_by_row(const Table &table, std::ostream &os,
+                                           std::vector<size_t> requested_row_ids);
+    static void write_ipac_subtable_by_row(const Table &table, std::ostream &os,
+                                           size_t start_row, size_t row_count);
+    static void write_single_ipac_record(const Table &table, std::ostream &os,
+                                         size_t row_idx);
+    static void write_consecutive_ipac_records(const Table &table, std::ostream &os,
+                                               size_t start_row, size_t row_count);
```
[the remark this answers is not in the document]
Comment (judith-ipac) [reply src/Ipac_Table_Writer.hxx]: Yes.  It seemed possible that some clients might just want records.  Angela wants the header also.

File: src/Ipac_Table_Writer/internals.cxx (hunk 6 above)
[the remark this answers is not in the document]
Comment (smonkewitz) [reply src/Ipac_Table_Writer/internals.cxx]: I'm pretty sure you moved it in one of the subsequent commits 😄

**Pull request merged**